vm/src/org/mozartoz/truffle/nodes/call/CallMethodNode.java

```diff
@@ -33,11 +45,52 @@ public static CallMethodNode create() {
 
 	static final RecordFactory OTHERWISE_MESSAGE_FACTORY = Arity.build("otherwise", 1L).createFactory();
 
-	@Specialization
+	@Specialization(guards = {
+			"self.getClazz() == cachedClazz",
+			"cachedLabel.equals(labelFromArgs(args))",
+			"cachedProc != null",
+	}, limit = "methodCache(3)")
+	protected Object callClassLabelIdentity(VirtualFrame frame, OzObject self, Object[] args,
+			@Cached("self.getClazz()") DynamicObject cachedClazz,
+			@Cached("labelFromArgs(args)") Object cachedLabel,
+			@Cached("methodFromArgs(self, args)") OzProc cachedProc,
+			@Cached("createDirectCallNode(cachedProc.callTarget)") DirectCallNode callNode) {
+		assert args.length == 1;
+		Object[] arguments = new Object[] { self, args[0] };
+		return callNode.call(frame, OzArguments.pack(cachedProc.declarationFrame, arguments));
+	}
+
+	@Specialization(guards = {
+			"cachedProc == methodFromArgs(self, args)",
+			"cachedProc != null",
+	}, contains = "callClassLabelIdentity", limit = "methodCache(3)")
+	protected Object callProcIdentity(VirtualFrame frame, OzObject self, Object[] args,
+			@Cached("methodFromArgs(self, args)") OzProc cachedProc,
+			@Cached("createDirectCallNode(cachedProc.callTarget)") DirectCallNode callNode) {
+		assert args.length == 1;
+		Object[] arguments = new Object[] { self, args[0] };
+		return callNode.call(frame, OzArguments.pack(cachedProc.declarationFrame, arguments));
+	}
+
+	@Specialization(guards = {
+			"self.getClazz() == cachedClazz",
+			"cachedLabel.equals(labelFromArgs(args))",
+			"methodNotFound",
+	}, limit = "methodCache(3)")
+	protected Object callOtherwise(VirtualFrame frame, OzObject self, Object[] args,
+			@Cached("self.getClazz()") DynamicObject cachedClazz,
+			@Cached("labelFromArgs(args)") Object cachedLabel,
+			@Cached("methodFromArgs(self, args) == null") boolean methodNotFound,
+			@Cached("getMethod(self, OTHERWISE)") OzProc otherwiseProc,
+			@Cached("createDirectCallNode(getMethod(self, OTHERWISE).callTarget)") DirectCallNode callNode) {
```

Comment:
Use otherwiseProc instead of `getMethod(self, OTHERWISE`. The general rule is cached expressions should use cached values or be guarded.